---
title: "Sample report for SARS-CoV-2 minipipeline"
output: 
  html_document:
    toc: true
params:
  mode: "none"
  fastp_table_stats: "none"
  fastp_table_stats_filter: "none"
  kraken_table: "none"
  flagstat_table: "none"
  fragment_size_table: "none"
  fragment_size_median_table: "none"
  coverage_table: "none"
  positive: "none"
  negative: "none"
  sample_cov: "none"
  president_results: "none"
  pangolin_results: "none"
  nextclade_results: "none"
  nextclade_version: "none"
  nextclade_dataset_version: "none"
  sc2rf_results: "none"
  vois_results: "none"
  cns_min_cov: "none"
  run_id: "none"
  pipeline_version: "none"
---

<style type="text/css">
.main-container {
max-width: 75%;
}
</style>

<style>
  .superwideimage{
      overflow-x:scroll;
      white-space: nowrap;
  }

  .superwideimage img{
    max-width: none;
  }
  
</style>

<style>
  .superhighimage{
      overflow:auto;
      height: 500px;
      width: 100%;
      margin-top: 10px;
      margin-bottom: 
  }

</style>

<style>
    div.scroll {
        width: 100%;
        height: 300px;
        overflow-x: hidden;
        overflow-y: auto;
        text-align: center;
        padding: 20px;
    }
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, error = FALSE)

library("data.table")
library("formattable")
library("ggplot2")
library("kableExtra")
library("plyr")
library("rjson")
library("dplyr")
library("DT")
library("xlsx")
```

```{r get_cmd_line_parameters}
dt.fastp_table_stats <- as.data.table(ldply(params$fastp_table_stats, fread))
dt.fastp_table_stats_filter <- as.data.table(ldply(params$fastp_table_stats_filter, fread))
if(params$kraken_table != 'none'){
  dt.kraken_table <- as.data.table(ldply(params$kraken_table, fread))
} else {
  print("No Kraken2 run.")
}
dt.flagstat_table <- as.data.table(ldply(params$flagstat_table, fread))
dt.fragment_size_table <- as.data.table(ldply(params$fragment_size_table, fread))
dt.fragment_size_median_table <- as.data.table(ldply(params$fragment_size_median_table, fread))
dt.coverage_table <- as.data.table(ldply(params$coverage_table, fread))
dt.positive <- as.data.table(ldply(params$positive, fread))
dt.negative <- as.data.table(ldply(params$negative, fread))
dt.sample_cov <- as.data.table(ldply(params$sample_cov, fread))
dt.president_results <- as.data.table(ldply(params$president_results, fread, sep='\t'))
dt.pangolin_results <- as.data.table(ldply(params$pangolin_results, fread))
dt.nextclade_results <- as.data.table(ldply(params$nextclade_results, fread, sep='\t'))
dt.sc2rf_results <- as.data.table(ldply(params$sc2rf_results, fread, sep=','))
names(dt.sc2rf_results) <- c('sample','examples','intermissions','breakpoints','regions')
if(params$vois_results != 'none'){
  dt.vois_results <- as.data.table(ldply(params$vois_results, fread))
}
```

```{r set_plot_width}
sample.count <- length(unique(dt.fastp_table_stats$sample))
# limit plot size to avoid oversized plots beyond limits of png size
if (sample.count > 100) {
  plot.width <- 50
} else if (sample.count < 10) {
  plot.width <- 10  
} else {
  plot.width <- sample.count * 0.5
}
```

```{r set_plot_height}
height <- ceiling(sample.count/3)
plot.height <- height * 1.9

# limit plot size to avoid oversized plots beyond limits of png size
if (plot.height > 50) {
  plot.height <- 50
}
```

```{r def_functions}
# modified color_bar to fix direction from "rtl" to "ltr"
f.color_bar <- function (color = "lightgray", fun = "proportion", 
    ...) 
{
    fun <- match.fun(fun)
    formattable::formatter("span", style = function(x) style(display = "inline-block", 
        direction = "ltr", `border-radius` = "4px", 
        `padding-right` = "2px", `background-color` = csscolor(color), 
        width = percent(fun(as.numeric(x), ...))))
}
```

## Sequencing run

```{r conditional_sentence, results='asis'}
if (params$run_id != 'none') {
  cat(paste0("Sequencing was performed in run: ", params$run_id, "\n"))
} else {
  cat(paste0("No ID for a sequencing run was submitted. \n"))
}
```


```{r conditional_table}
# if (nrow(dt.sample_cov_negative) > 0) {
#     dt.sample_cov_tmp <- dt.sample_cov
#     dt.sample_cov_tmp$genome.coverage <- round(dt.sample_cov_tmp$genome.coverage, 2)

#     # add a row with all 0 and 1 to make colour scaling reproducible
#     df.tmp <- data.frame("sample" = c(0,1), 
#                          "genome.length" = c(0,1),
#                          "covered.bases" = c(0,1), 
#                          "genome.coverage" = c(0,1)
#                          )
#     dt.sample_cov_tmp <- rbind(dt.sample_cov_tmp, df.tmp)
#     rm(df.tmp)
#     # highlight important rows
#     dt.sample_cov_tmp$sample <- cell_spec(dt.sample_cov_tmp$sample, color = ifelse(dt.sample_cov_tmp$genome.coverage > 0.2, "red", "black"))
#     dt.sample_cov_tmp$genome.coverage <- color_tile("white", "orange")(dt.sample_cov_tmp$genome.coverage)
#     #remove added lines
#     dt.sample_cov_tmp <- head(dt.sample_cov_tmp, n = -2)
    
#     datatable(dt.sample_cov_tmp[,c("sample", "covered.bases", "genome.coverage")], 
#         colnames = c("sample", "ref.coverage [bp]", "ref.coverage [fraction]"), 
#         caption = "Reference genome coverage in base pairs and as fraction of the total length. Bases with more than 10x sequencing depth are counted as covered. Negative control's names are highlighted in red if more than 20% of reference genome is covered, increasing coverage fraction is coloured orange (scaled from 0 to 1).",
#         options = list(order = list(list(0, 'asc'))),
#         escape = F)

#     rm(dt.sample_cov_tmp)
# }
```

```{r overview_table}
dt.overview <- merge(x=dt.fastp_table_stats[,c('sample', "reads.before.clip")], y=dt.flagstat_table[,c('sample', "input")], by='sample')

if(params$kraken_table != 'none'){
  dt.overview <- merge(x=dt.overview, y=dt.kraken_table, by='sample')
}

dt.president_results$N_percent <- (dt.president_results$N_bases / dt.president_results$length_query) * 100
dt.president_results <- dt.president_results[,c("query_name", "qc_all_valid", "ACGT Nucleotide identity", "N_bases", "N_percent", "Mismatches")]
colnames(dt.president_results)[1] <- 'sample'
dt.overview <- merge(x=dt.overview, y=dt.president_results, by='sample')

dt.pangolin_results_tmp <- dt.pangolin_results[,c("taxon", "lineage", "conflict", "scorpio_call", "scorpio_conflict", "qc_status")]
dt.overview <- merge(x=dt.overview, y=dt.pangolin_results_tmp, by.x='sample', by.y='taxon')

dt.overview <- merge(x=dt.overview, y=dt.nextclade_results[,c("seqName", "clade", "Nextclade_pango", "substitutions", "aaSubstitutions", "deletions", "aaDeletions",	"insertions", "aaInsertions", "frameShifts", "privateNucMutations.reversionSubstitutions",	"privateNucMutations.labeledSubstitutions",	"privateNucMutations.unlabeledSubstitutions","qc.privateMutations.status")], by.x='sample', by.y='seqName')

dt.sc2rf_results$recomb <- ifelse(dt.sc2rf_results$intermissions <= 3 & dt.sc2rf_results$breakpoints > 0 & dt.sc2rf_results$breakpoints <= 2, 'yes', 'no')
dt.overview <- merge(x=dt.overview, y=dt.sc2rf_results[,c("sample", "recomb", "regions")], by='sample')

if(params$kraken_table != 'none') {
  colnames(dt.overview) <- c('sample', '# total reads', '# filtered reads', 'kraken2_ratio_human', 'kraken2_ratio_sarscov2', 'kraken2_ratio_unclassified', 'kraken2_numreads_human', 'kraken2_numreads_sarscov2', 'kraken2_numreads_unclassified', 'president_QC_pass', 'president_identity', 'president_numN', 'president_percentN', 'president_mismatches', 'pangolin_lineage', 'pangolin_conflict', 'scorpio_constellation', 'scorpio_conflict', 'pangolin_status', 'nextclade_clade', 'nextclade_pango', 'nextclade_mutations_nt', 'nextclade_mutations', 'nextclade_deletions_nt', 'nextclade_deletions', 'nextclade_insertions_nt', 'nextclade_insertions', 'nextclade_frameshifts', 'nextclade_private_nt_mutation_reversionSubstitutions', 'nextclade_private_nt_mutation_labeledSubstitutions', 'nextclade_private_nt_mutation_unlabeledSubstitutions', 'nextclade_privateMutations_status', 'sc2rf_potential_recombinant', 'sc2rf_regions')
} else {
  colnames(dt.overview) <- c('sample', '# total reads', '# filtered reads', 'president_QC_pass', 'president_identity', 'president_numN', 'president_percentN', 'president_mismatches', 'pangolin_lineage', 'pangolin_conflict', 'scorpio_constellation', 'scorpio_conflict', 'pangolin_status', 'nextclade_clade', 'nextclade_pango', 'nextclade_mutations_nt', 'nextclade_mutations', 'nextclade_deletions_nt', 'nextclade_deletions', 'nextclade_insertions_nt', 'nextclade_insertions', 'nextclade_frameshifts', 'nextclade_private_nt_mutation_reversionSubstitutions', 'nextclade_private_nt_mutation_labeledSubstitutions', 'nextclade_private_nt_mutation_unlabeledSubstitutions', 'nextclade_privateMutations_status', 'sc2rf_potential_recombinant', 'sc2rf_regions')
}

# dt.overview[dt.overview == ""] <- "n/a"
dt.overview[is.na(dt.overview)] <- ""

write.csv( x=dt.overview, 
            row.names = FALSE,
            file = file.path("report_datatable.csv"))
write.xlsx( x=dt.overview, 
            row.names = FALSE,
            file = file.path("report_datatable.xlsx"))
```

## Short summary
```{r conditional_warning, results='asis'}
dt.sample_cov_negative <- dt.sample_cov[grepl("NK|Empty", dt.sample_cov$sample, ignore.case = TRUE) & dt.sample_cov$genome.coverage >= 0.2, ]

if (nrow(dt.sample_cov_negative) > 0) {
    cat("<p style=\"color:red\">Warning: Samples identified automatically as negative controls show unusual large coverage of the reference genome (more than 20%). Please check, if this is expected for the samples.</p>\n")
}
```

```{r overview}
dt.overview$president_QC_pass <- ifelse(dt.overview$president_QC_pass == TRUE,
                                    cell_spec('PASS', background = "lightgreen", align = "right"),
                                    cell_spec('FAIL', background = "red", align = "right"))
dt.overview$nextclade_privateMutations_status <- ifelse(dt.overview$nextclade_privateMutations_status == 'good',
                                    cell_spec(dt.overview$nextclade_privateMutations_status, background = "lightgreen", align = "right"),
                                    ifelse(dt.overview$nextclade_privateMutations_status == 'mediocre',
                                    cell_spec(dt.overview$nextclade_privateMutations_status, background = "orange", align = "right"),
                                    ifelse(dt.overview$nextclade_privateMutations_status == '',
                                    dt.overview$nextclade_privateMutations_status, 
                                    cell_spec(dt.overview$nextclade_privateMutations_status, background = "red", align = "right"))))
dt.overview$sc2rf_potential_recombinant <- ifelse(dt.overview$sc2rf_potential_recombinant == 'yes',
                                    cell_spec(dt.overview$sc2rf_potential_recombinant, background = "orange", align = "right"),
                                    dt.overview$sc2rf_potential_recombinant)

if(params$kraken_table != 'none'){
  col <- c("sample", '# total reads', '# filtered reads', "kraken2_ratio_sarscov2", "president_QC_pass", "president_identity", "president_percentN", "pangolin_lineage","scorpio_constellation", "nextclade_frameshifts","nextclade_privateMutations_status", 'sc2rf_potential_recombinant')
  col_names <- c("Sample", '# total reads', '# filtered reads', "Severe acute respiratory syndrome coronavirus 2 (ratio)", "Genome QC", "%Identity", "%N", "Linage", "Scorpio call", "Frameshifts", "Private mutations QC", 'Potential recombinant')
} else {
  col <- c("sample", '# total reads', '# filtered reads', "president_QC_pass", "president_identity", "president_percentN", "pangolin_lineage", "scorpio_constellation", "nextclade_frameshifts","nextclade_privateMutations_status", 'sc2rf_potential_recombinant')
  col_names <- c("Sample", '# total reads', '# filtered reads', "Genome QC", "%Identity", "%N", "Linage", "Scorpio call", "Frameshifts", "Private mutations QC", 'Potential recombinant')
}

datatable(dt.overview[, col, with=FALSE], 
      colnames = col_names,
      rownames = FALSE,
      caption = "Short overview over the pipeline run. Please see below or pipeline results for more extensive output.",
      options = list(scrollX = TRUE, order = list(list(0, 'asc')), columnDefs = list(list(targets="_all", render = JS("function(data, type, row, meta) {return data === null || data === '' ? 'n/a' : data;}")))),
      escape = FALSE)
```

## Trimming statistics

Trimming and clipping is performed using fastp.

### Read counts

Raw reads were subjected to adapter clipping.

```{r table_trimming, fig.cap="Counts of reads before and after clipping as well as base count of >=Q30."}
dt.table <- dt.fastp_table_stats

# add coloured bar charts to table
dt.table$reads.before.clip <- f.color_bar("lightgreen")(dt.table$reads.before.clip)
dt.table$reads.after.clip <- f.color_bar("lightgreen")(dt.table$reads.after.clip)

datatable(dt.table,
    rownames = FALSE,
    colnames = c("Sample", "Reads before clip", "Reads after clip", "% passed", "Q30 before clip", "Q30 after clip"),
    caption = "Counts of reads before and after clipping as well as base count of >=Q30. Read counts before and after clipping augmented with bar chart for visual aid.",
    options = list(order = list(list(0, 'asc'))),
    escape = F) %>%
    formatRound(columns=c("ratio.clip", "q30.before.clip", "q30.after.clip"), digits=2)

rm(dt.table)
```

<div class="superwideimage">

```{r plot_trimming, fig.width=plot.width, fig.cap="Read count before and past adapter clipping."}
df.plot <- melt(dt.fastp_table_stats,
                id.vars = "sample",
                measure.vars = c("reads.before.clip", "reads.after.clip"))

ggplot(data = df.plot,
       aes(sample, value, fill = variable)) +
    geom_bar(stat = "identity", position = "dodge") +
    labs(title = "read counts before & after trimming",
         x = "sample",
         y = "count",
         fill = "clipping") +
    scale_fill_manual(values = c("#DD4444", "#4444DD"), labels = c("raw", "clipped")) +
    theme(plot.title = element_text(hjust = 0.5), 
          axis.text.x = element_text(angle = 45, hjust = 1),
          legend.position = "bottom")
  
rm(df.plot)
```

</div>

### Filtering details

Reads were subjected to adapter clipping and quality trimming. 
The latter includes low complexity filtering as well as classical Phred based quality filtering. 

```{r table_filtering, fig.cap="Counts of reads after clipping and counts of reads removed by several filters."}

dt.fastp_table_stats_filter$passed_filter <- f.color_bar("lightgreen")(dt.fastp_table_stats_filter$passed_filter)

datatable(dt.fastp_table_stats_filter,
    rownames = FALSE,
    colnames = c("Sample", "# reads passed filter", "Low quality reads", "High #N", "Low complexity reads", "Too short reads"),
    caption = "Counts of reads after clipping and counts of reads removed by several filters. Passed filter read counts augmented with bar chart for visual aid.",
    options = list(order = list(list(0, 'asc'))),
    escape = FALSE)
```

### Species filtering

Reads can be filtered against a defined taxonomical ID predefined at the start of the pipeline run. 
If this parameter was set and Kraken2 classification was successful, filtering data will be shown here.

```{r table_kraken, fig.cap="Read counts after species binning using Kraken2."}
if (params$kraken_table != 'none') {
  datatable(dt.kraken_table, 
      rownames = FALSE,
      caption = "Amount of reads assigned to selected species. Ratios are coloured to highlight high unclassified content (scaled 0-100%)", escape = F,
      options = list(order = list(list(0, 'asc')), columnDefs = list(list(targets="_all", render = JS("function(data, type, row, meta) {return data === 'NA' ? '0' : data;}")))))
      #  %>%
      # formatRound(columns=c("Homo sapiens (ratio)","Severe acute respiratory syndrome coronavirus 2 (ratio)","unclassified (ratio)"), digits=2)
} else {
  print("No Kraken2 run.")
}
```

## Mapping statistics

Species classification of the reads may have been performed.
If this was done, reads assigned to the defined species were submitted to genome mapping. If classification was omitted reads after fastp filtering were used for mapping.

Reads were mapped to the reference genome using BWA. 

DNA fragments are counted twice, if paired end sequencing was employed.

```{r table_flagstat}
dt.flagstat_table$input <- f.color_bar("lightgreen")(dt.flagstat_table$input)
dt.flagstat_table$mapped <- f.color_bar("lightgreen")(dt.flagstat_table$mapped)

datatable(dt.flagstat_table, 
    rownames = FALSE,
    colnames = c("Sample", "Total reads", "Reads mapped", "Mapping ratio"),
    caption = "Amount of reads subjected to reference genome mapping (after the optional Kraken filtering) and the respective numbers of mapped reads. In paired end sequencing each DNA fragment is counted twice. Input and mapped read counts augmented with bar chart for visual aid.",
    options = list(order = list(list(0, 'asc'))),
    escape = F) %>%
    formatRound(columns=c("mapping.rate"), digits=3)
```

`r if(params$mode == "paired"){"
## Fragment sizes

Fragment sizes were determined by mapping reads reference genome and extracting the fragment size from the respective bam file.
Data of forward and reverse mapping reads were used, which leads to a duplicated entry for each paired end mapping.
The log value of the median is given for comparison with the values shown in the plots.

### Fragment size table
"}`

```{r table_fragsizes, fig.cap="Median fragment size determined by the distance of reads after mapping to reference genome."}
if(params$mode == "paired"){
  # conditional formatting for fragsizes not between 90 and 110
  dt.fragment_size_median_table$median.fragsize <- ifelse(dt.fragment_size_median_table$median.fragsize >= 110 | dt.fragment_size_median_table$median.fragsize <= 90,
                                                cell_spec(dt.fragment_size_median_table$median.fragsize, background = "orange", align = "right"),
                                                dt.fragment_size_median_table$median.fragsize)

  datatable(dt.fragment_size_median_table,
      rownames = FALSE,
      colnames = c("Sample", "Fragment size (median)", "Fragment size (stdev)"),
      caption = "Median fragment size determined by the distance of reads after mapping to reference genome. Read distance not between 90 and 110 bp are highlighted in orange.",
      options = list(order = list(list(0, 'asc'))),
      escape = FALSE) %>%
      formatRound(columns=c("sd.fragsize"), digits=2)
}
```

`r if(params$mode == "paired"){"
### Fragment size plot

Plotting the fragment sizes usually shows a distribution with multiple peaks. Ideally just one peak dominates the plot at around 100bp.
"}`
<div class=superhighimage>

```{r plot_fragsizes, fig.width=10, fig.height=plot.height, fig.cap="Fragment size distribution determined by the distance of reads after mapping to reference genome.", error=T}
if(params$mode == "paired"){
  tryCatch({
      ggplot(dt.fragment_size_table, aes(fragsize.abs, colour = sample, fill = sample)) +
      geom_density() +
      facet_wrap(~sample, ncol = 4, scales = "free_y") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            legend.position = "none",
            plot.title = element_text(hjust = 0.5)) +
      scale_x_continuous(trans = "log10") +
      labs(title = "Fragment size histogram",
          x = "fragment size")
      }, error = function(e) {
          message("some error")
      })
}
```

</div>

## Coverage distribution

### Sequence depth distribution on reference genome

Sequence depth was calculated at each position and plotted. The aim for positive samples is an evenly distributed high sequence depth.

<div class=superhighimage>

```{r plot_coverage, fig.width=10, fig.height=plot.height, fig.cap="Sequence depth distribution on SARS-CoV-2 genome. Sequence depth is shown versus the position of the base in the genome. Please be aware of the varying y axis scaling."}

if (sum(dt.coverage_table$depth)>0) {
    # only apply log10 scaling, if any depth value is >0
    ggplot(dt.coverage_table[dt.coverage_table$mid.bin == 50,], aes(position, depth, colour = sample)) +
        geom_point(size = 0.1) +
        labs(title = "coverage distribution") +
        theme(plot.title = element_text(hjust = 0.5), 
          axis.text.x = element_text(angle = 45, hjust = 0.5),
          legend.position = "none") +
        scale_y_continuous(trans = "log10") +
        facet_wrap(~sample, ncol = 3, scales = "free_y")
} else {
    ggplot(dt.coverage_table[dt.coverage_table$mid.bin == 50,], aes(position, depth, colour = sample)) +
        geom_point(size = 0.1) +
        labs(title = "coverage distribution") +
        theme(plot.title = element_text(hjust = 0.5), 
          axis.text.x = element_text(angle = 45, hjust = 0.5),
          legend.position = "none") +
        facet_wrap(~sample, ncol = 3, scales = "free_y")
}
```

</div>

### Accumulated sequence depth

Sequence depth on the reference genome is visualized as histograms binning bases of identical coverage.

<div class=scroll>

```{r plot_coverage_histogram, fig.width=10, fig.height=plot.height, fig.cap="Sequence depth distribution on SARS-CoV-2 genome. The amount of bases is shown dependend on the respective sequence depth. Please watch out for the different axis scaling."}

ggplot(dt.coverage_table[dt.coverage_table$mid.bin == 50,], aes(depth, fill = sample)) +
    geom_histogram(binwidth = 50) +
    facet_wrap(~sample, ncol = 1, scales = "free") +  # ncol > 1 results in plots that are displayed incorrectly
    labs(title = "coverage histogram") +
    theme(plot.title = element_text(hjust = 0.5),
          legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 0.5)) +
    facet_wrap(~sample, ncol = 2, scales = "free_y")
```

</div>

### Tabular view of coverage distribution

```{r table_coverage}
df.tmp <- data.frame("sample" = c(0,1), 
                     "genome.length" = c(0,1),
                     "covered.bases" = c(0,1), 
                     "genome.coverage" = c(0,1),
                     "DP.median" = c(0,0),
                     "DP.mean" = c(0,0)
)
dt.sample_cov_table <- rbind(dt.sample_cov, df.tmp)
rm(df.tmp)

# conditional formatting for reference coverage >95%
dt.sample_cov_table$genome.coverage <- round(dt.sample_cov_table$genome.coverage, 2)
dt.sample_cov_table$genome.coverage <- ifelse(dt.sample_cov_table$genome.coverage >= 0.95,
                                    cell_spec(dt.sample_cov_table$genome.coverage, background = "lightgreen", align = "right"),
                                    dt.sample_cov_table$genome.coverage)
# coverage bar
dt.sample_cov_table$genome.length <- f.color_bar("lightgreen")(dt.sample_cov_table$genome.length)

# highlight negative controls
dt.sample_cov_table$sample <- ifelse(grepl(pattern = "NK", dt.sample_cov_table$sample),
                           cell_spec(dt.sample_cov_table$sample, color = "red"),
                           dt.sample_cov_table$sample)

#remove added lines
dt.sample_cov_table <- head(dt.sample_cov_table, n = -2)
datatable(dt.sample_cov_table[,c("sample", "covered.bases", "genome.coverage", "DP.median", "DP.mean")], 
        rownames = FALSE,
        colnames = c("Sample", "Coverage [bp]", "Coverage [fraction]", "Depth median", "Depth mean"), 
        caption = paste0("Genome coverage values shown as coverage of the reference genome and sequencing depth. Bases with more than ", toString(params$cns_min_cov),"x sequencing depth are counted as covered. Sample names of negative controls are highlighted in red and reference genome coverage fraction >=95% is highlighted in green. Negative controls with more than 20% reference genome coverage could indicate contamination."),
        options = list(order = list(list(0, 'asc'))),
        escape = FALSE) %>%
    formatRound(columns=c("DP.mean"), digits=2)
rm(dt.sample_cov_table)
```

## Genome quality

Quality statistics by [PRESIDENT](https://gitlab.com/RKIBioinformaticsPipelines/president) (PaiRwisE Sequence IDENtiTy) with identity of ACGT bases of 90% and maximal 5% Ns in the query/sample compared to the reference genome (see RKI genome submission requirements: [PDF](https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/DESH/Qualitaetskriterien.pdf?__blob=publicationFile) in German).

```{r president}
# dt.president_results <- dt.president_results[,c("query_name","ACGT Nucleotide identity","ACGT Nucleotide identity (ignoring Ns)","ACGT Nucleotide identity (ignoring non-ACGTNs)","qc_all_valid","qc_is_empty_query","qc_post_align_pass_threshold","qc_post_aligned","qc_post_aligned_all_valid","qc_valid_length","qc_valid_nucleotides","qc_valid_pass_nthreshold","acgt_bases","iupac_bases","non_iupac_bases","N_bases","length_query","length_reference","LongestNGap","Matches","Mismatches")]

dt.president_results$qc_all_valid <- ifelse(dt.president_results$qc_all_valid == TRUE,
                                    cell_spec('PASS', background = "lightgreen", align = "right"),
                                    cell_spec('FAIL', background = "red", align = "right"))
datatable(dt.president_results, 
      colnames = c("Sample", "QC status", "ACGT Nucleotide identity", "#N", "%N", "Mismatches"),
      rownames = FALSE,
      caption = "Quality statistics by PRESIDENT.",
      options = list(scrollX = TRUE, order = list(list(0, 'asc')), columnDefs = list(list(targets="_all", render = JS("function(data, type, row, meta) {return data === null ? 'NA' : data;}")))),
      escape = FALSE)
```

## Lineage overview

Covpipe runs a preliminary lineage assignment with Pangolin (https://github.com/cov-lineages/pangolin). Note that there are several lineage assignment schemes that need to get run in separate analyses steps.

Versions: 

- pangolin: `r dt.pangolin_results$pangolin_version[1]`
- pangolin-data: `r dt.pangolin_results$version[1]`
- scorpio: `r dt.pangolin_results$scorpio_version[1]` with constellation `r dt.pangolin_results$constellation_version[1]`

```{r lineage_assignment}
dt.pangolin_results$qc_status <- ifelse(dt.pangolin_results$qc_status == "pass",
                                    cell_spec('PASS', background = "lightgreen", align = "right"),
                                    cell_spec('FAIL', background = "red", align = "right"))
datatable(dt.pangolin_results[,c("taxon", "lineage", "conflict", "scorpio_call", "scorpio_conflict", "qc_status")], 
      colnames = c("Sample", "Lineage", "Conflict", "Scorpio call", "Scorpio conflict", "QC status"),
      rownames = FALSE,
      caption = "Preliminary lineage assignment using Pangolin.",
      options = list(scrollX = TRUE, order = list(list(0, 'asc')), columnDefs = list(list(targets="_all", render = JS("function(data, type, row, meta) {return data === null ? 'NA' : data;}")))),
      escape = FALSE)
```

## Mutations
Clade, mutations, deletions, insertions and frameshifts were determined with Nextclade (https://clades.nextstrain.org/).

Versions: `r params$nextclade_version` with dataset tag `r params$nextclade_dataset_version`
```{r nextclade}
datatable(dt.nextclade_results[,c("seqName", "aaSubstitutions", "aaDeletions", "aaInsertions", "frameShifts")], 
      colnames = c("Sample", " AA substitutions", "AA deletions", "AA insertions", "Frameshifts"),
      rownames = FALSE,
      caption = "Mutation profile",
      options = list(scrollX = TRUE, order = list(list(0, 'asc')), autoWidth = FALSE, columnDefs = list(list(targets="_all", render = JS("function(data, type, row, meta) {return data === null ? '-' : data;}")))),
      escape = FALSE)
```

## Variation sites of interest

Covpipe examines variation sites of interest. 

```{r voi_table}
if (params$vois_results != 'none') {
  datatable(dt.vois_results, 
        rownames = FALSE,
        caption = paste0("Variation sites of interest and their occurrence in the input samples. Low coverage is below ", toString(params$cns_min_cov), ". If a variant is found with an other ALT in the sample, this variation is stated."),
        escape = FALSE)
} else {
  print("No variants of interest.")
}
```

## Pipeline version
```{r conditional_sentence_pipeline_version, results='asis'}
if (params$pipeline_version != 'none') {
   cat(paste0("This report was generated using covpipe pipeline version", params$pipeline_version, ".\n"))
} else {
   cat("No version available - local execution.\n")
}
```
